src/Interpreters/InterpreterSystemQuery.cpp

```diff
@@ -583,14 +600,22 @@ void InterpreterSystemQuery::dropReplica(ASTSystemQuery & query)
     }
     else if (query.is_drop_whole_replica)
     {
-        getContext()->checkAccess(AccessType::SYSTEM_DROP_REPLICA);
         auto databases = DatabaseCatalog::instance().getDatabases();
+        auto access = getContext()->getAccess();
+        bool access_is_granted_globally = access->isGranted(AccessType::SYSTEM_DROP_REPLICA);
 
         for (auto & elem : databases)
         {
             DatabasePtr & database = elem.second;
             for (auto iterator = database->getTablesIterator(getContext()); iterator->isValid(); iterator->next())
+            {
+                if (!access_is_granted_globally && !access->isGranted(AccessType::SYSTEM_DROP_REPLICA, elem.first, iterator->name()))
```

Comment:
@vitlibar not sure if this place need to be optimized, but it seems that we can move this check `!access_is_granted_globally` outside of loop, and just make log something that access it not granted globally to all entities. Same can be applied to similar loop above.

Reply:
If the access is not granted globally for all databases and tables it can be still granted for some tables. `isGranted()` also writes the result of checking access to the logs no matter it's positive (granted) or negative (denied), so I think perhaps there are enough logs already.

Reply:
Agreed.